pymongo/auth.py

```diff
@@ -557,15 +575,15 @@ def speculate_command(self):
 
 
 class _X509Context(_AuthContext):
-    def speculate_command(self):
+    def speculate_command(self) -> Optional[SON[str, Any]]:
         cmd = SON([("authenticate", 1), ("mechanism", "MONGODB-X509")])
         if self.credentials.username is not None:
             cmd["user"] = self.credentials.username
         return cmd
 
 
 class _OIDCContext(_AuthContext):
-    def speculate_command(self):
+    def speculate_command(self) -> Optional[SON[str, Any]]:
```

Comment:
When would speculate_command here return None? @blink1073 is there a way to avoid that? It's unexpected because pool.py doesn't handle this case:
```
            if auth_ctx:
                cmd["speculativeAuthenticate"] = auth_ctx.speculate_command()
```

Reply:
The parent class `_AuthContext`'s `speculate_command` simply raises a `NotImplementedError`, which causes mypy to complain if the subclass `speculate_command` does not have an `Optional` return type.

Reply:
The problem is not quite the NotImplementedError in the parent class, that part is fine because exceptions don't affect the return type. The problem is that _OIDCContext's speculate_command implementation can return None which is unexpected:
```python
        cmd = authenticator.auth_start_cmd(False)
        if cmd is None:
            return None # <- Uh?
        ...
```

Reply:
Ahh, got it. 

Looks like the actual code that can return None is here, in `auth_odic.py`'s `get_current_token`:

```        
if not use_callbacks and not current_valid_token:
	return None
```

Reply:
Hmm I see. I suspect we do actually need these changes to make the return Optional, but we need Steve's to confirm since he owns the OIDC feature. If so we also need to update pool.py to handle a None return.

Reply:
Yes, we do not want to send a speculate command if there is no current token, and we don't want to call a callback either, so Optional is appropriate.

Reply:
Do we want a separate ticket to handle the pool.py updates?

Reply:
I think we can do it in this PR. If speculate_command() returns None then we should skip adding the "speculativeAuthenticate" field.